Nfiann-arguments (#7128)
fishtown-analytics/docs.getdbt.com: website/docs/reference/resource-properties/arguments.md

You can validate your macro `arguments` using the [`validate_macro_args`](/reference/global-configs/behavior-changes#macro-argument-validation) flag:

- If the flag is set to `False` (default), dbt will continue to permit any value for `type` and `name`.
- If flag is set to `True` (opt-in), dbt will raise a warning if the argument names you've added in YAML don't match the argument names you have in your macro or if the argument types aren't valid according to the [supported types](/reference/resource-properties/arguments#supported-types). Additionally, if no argument names are documented in YAML, dbt will infer them based on what you have in the macro and include them in the [manifest.json](/reference/artifacts/manifest-json) file.

## type

<VersionBlock lastVersion="1.9">

The data type of your argument. This is only used for documentation purposes — there are no restrictions on the values you can use here.

</VersionBlock>
<VersionBlock firstVersion="1.10">

> The data type of your argument. Setting [`validate_macro_args`](/reference/resource-properties/arguments#supported-types) to `true` ensures that documented macro argument names match those in the macro definition and validates their types against the [supported types](#supported-types). When set to `false`, `type` is only used for documentation purposes and there are no restrictions on the values you can specify.

Reviewer: ```suggestion
The data type of your argument. Setting [`validate_macro_args`](/reference/global-configs/behavior-changes#macro-argument-validation) to `true` ensures that documented macro argument names match those in the macro definition and validates their types against the [supported types](#supported-types). When set to `false`, `type` is only used for documentation purposes and there are no restrictions on the values you can specify.
```

---
title: arguments
sidebar_label: "arguments"
id: arguments
---

import MacroArgsNote from '/snippets/_validate-macro-args.md';

<File name='macros/<filename>.yml'>

```yml

version: 2

macros:
  - name: <macro name>
    arguments:
      - name: <arg name>
        type: <string>
        description: <markdown_string>

```

</File>

## Definition

The `arguments` property is used to define the parameters that a macro can accept. Each argument can have a `name`, `type`, and `description`.  You add `arguments` to a [macro property](/reference/macro-properties), which helps in documenting the macro and understanding what inputs it requires.

> You can validate your macro `arguments` using the [`validate_macro_args`](/reference/global-configs/behavior-changes#macro-argument-validation) flag. 

Reviewer: ```suggestion
```
```suggestion
You can validate your macro `arguments` using the [`validate_macro_args`](/reference/global-configs/behavior-changes#macro-argument-validation) flag:
```

- If the flag is set to `False` (default), dbt will continue to permit any value for `type` and `name`.

> - If flag is set to `True` (opt-in), dbt will raise a warning if the argument names you've added in YAML don't match the argument names you have in your macro or if the argument types aren't valid according to the [supported types](/reference/global-configs/behavior-changes#supported-types).

Reviewer 2: this and any similar link needs updating 

```suggestion
- If flag is set to `True` (opt-in), dbt will raise a warning if the argument names you've added in YAML don't match the argument names you have in your macro or if the argument types aren't valid according to the [supported types](#supported-types-for-macro-argument-validation).
```

> - If flag is set to `True` (opt-in), dbt will raise a warning if the argument names you've added in YAML don't match the argument names you have in your macro or if the argument types aren't valid according to the [supported types](/reference/resource-properties/arguments#supported-types).
> 
> If no argument names are documented in YAML, dbt will infer them based on what you have in the macro and include them in the [manifest.json](/reference/artifacts/manifest-json) file.

Reviewer 3: ```suggestion
- If flag is set to `True` (opt-in), dbt will raise a warning if the argument names you've added in YAML don't match the argument names you have in your macro or if the argument types aren't valid according to the [supported types](/reference/resource-properties/arguments#supported-types). Additionally, if no argument names are documented in YAML, dbt will infer them based on what you have in the macro and include them in the [manifest.json](/reference/artifacts/manifest-json) file.
```

## Macro properties

> Macro properties can be declared in any `properties.yml` file. Macro properties are "special properties" in that you can't configure them in the dbt_project.yml file or using config() blocks. Refer to [Configs and properties](/reference/define-properties#which-properties-are-not-also-configs) for more information.

Reviewer: I would also add backticks here:
```suggestion
Macro properties can be declared in any `properties.yml` file. Macro properties are "special properties" in that you can't configure them in the dbt_project.yml file or using `config()` blocks. Refer to [Configs and properties](/reference/define-properties#which-properties-are-not-also-configs) for more information.
```

<VersionBlock lastVersion="1.9">

The data type of your argument. This is only used for documentation purposes — there are no restrictions on the values you can use here.

</VersionBlock>
<VersionBlock firstVersion="1.10">

> The data type of your argument. Unless you use the [`validate_macro_args`](/reference/resource-properties/arguments#supported-types) flag, `type` is only used for documentation purposes — there are no restrictions on the values you can use here.

Reviewer 3: I would flip this sentence and say "when you set `validate_macros_args` to True ..."

> The data type of your argument. Unless you use the [`validate_macro_args`](/reference/resource-properties/arguments#supported-types) flag, `type` is only used for documentation purposes — there are no restrictions on the values you can use here.

Reviewer: ```suggestion
The data type of your argument. Setting [`validate_macro_args`](/reference/resource-properties/arguments#supported-types) to `true` ensures that documented macro argument names match those in the macro definition and validates their types against the [supported types](#supported-types). When set to `false`, `type` is only used for documentation purposes and there are no restrictions on the values you can specify.
```

</VersionBlock>

<MacroArgsNote />

<File name='macros/<filename>.yml'>

```yml
version: 2

macros:
  - name: <macro name>
    arguments:
      - name: <arg name>
        type: <string>

```

</File>

> ### Supported types for macro argument validation

Reviewer 2: this might be just me but i feel this header can be shortened to just 'supported types' since it's nested under types and the intro text explains what its for. if you do change it, anything that links to this section will need updating

> ## type

Reviewer: Can you also do these tasks I didn't do so I wouldn't create merge conflicts with your PR:

[ ] move the supported types from [here](https://github.com/dbt-labs/docs.getdbt.com/pull/7112/files#diff-1cf86e82c385dd5fa31e85ba75cc93d6170a661d06b082bbcb05c7a5a00a8451R261-R274) to this type section? Thank you!! 🙇🏻‍♀️ 

[x] This also involves finding all the previous links to `/reference/global-configs/behavior-changes#supported-types` and changing them to this new location.

From dbt Core v1.10, when you use the [`validate_macro_args`](/reference/global-configs/behavior-changes#macro-argument-validation) flag, dbt supports the following types for macro arguments:

- `string` or `str`
- `boolean` or `bool`
- `integer` or `int`
- `float`
- `any`
- `list[<Type>]`, for example, `list[string]`
- `dict[<Type>, <Type>]`, for example, `dict[str, list[int]]`
- `optional[<Type>]`, for example, `optional[integer]`
- [`relation`](/reference/dbt-classes#relation)
- [`column`](/reference/dbt-classes#column)

Note that the types follow a Python-like style but are used for documentation and validation only. They are not Python types.

## Examples

> ### Document a macro

Reviewer 3: ```suggestion
```

---
title: arguments
sidebar_label: "arguments"
id: arguments
---

import MacroArgsNote from '/snippets/_validate-macro-args.md';

<File name='macros/<filename>.yml'>

```yml

version: 2

macros:
  - name: <macro name>
    arguments:
      - name: <arg name>
        type: <string>
        description: <markdown_string>

```

</File>

## Definition

The `arguments` property is used to define the parameters that a macro can accept. Each argument can have a `name`, `type`, and `description`.  You add `arguments` to a [macro property](/reference/macro-properties), which helps in documenting the macro and understanding what inputs it requires.

> 

Reviewer: ```suggestion
```

> You can use a [behavior flag](/reference/global-configs/behavior-changes#behavior-change-flags) in v1.10; dbt will validate whether the `name` or `type` matches the marcro arguments in YAML.

Reviewer: I'm a little unclear on what this paragraph is communicating. We should probably mention which behavior flag you can use. Or are we saying that there is a behavior flag introduced in 1.10 that will affect macro arguments?

> `arguments` can be defined in a [`properties.yml`](/reference/resource-properties/arguments) file but they are not enforced by dbt. The file can be named anything and placed in the [`macro-paths`](/reference/project-configs/macro-paths) directory (defaults to `["macros"]`).

Author: ```suggestion
`arguments` can be defined in a [`properties.yml`](/reference/resource-properties/arguments#macro-properties) file but they are not enforced by dbt. The file can be named anything and placed in the [`macro-paths`](/reference/project-configs/macro-paths) directory (defaults to `["macros"]`).
```

> `arguments` can be defined in a [`properties.yml`](/reference/resource-properties/arguments#macro-properties) file but they are not enforced by dbt. The file can be named anything and placed in the [`macro-paths`](/reference/project-configs/macro-paths) directory (defaults to `["macros"]`).

Reviewer: 1. Can you make these sentences active? For example, "You can define `arguments`..." and " You can name this file..."
2. I wonder if this paragraph should move up to before the behavior flag paragraph? It might flow better.

## Macro properties

> import PropsCallout from '/snippets/_config-prop-callout.md';

Reviewer: This can be removed if the next comment is correct.

> Macro properties can be declared in any `properties.yml` file. <PropsCallout title={frontMatter.title}/> 

Reviewer: Is  `<PropsCallout title={frontMatter.title}/>` in the wrong place? It's using "arguments" instead of "macro properties" and arguments is lower case instead of capitalized.

Alternatively, you could use <PropsCallout title=Macroproperties/> (not sure if it needs quotes or not.)

Author: Hiya @runleonarun 

I think I know what happened with this one (I'll do my best to explain lol)
 
Initially "Macro properties" was a standalone page.
Grace advised this page could be added to the arguments page.
I copied and pasted the contents of Macro properties to the `arguments' page. The "<PropsCallout title={frontMatter.title}/> " pulled in the page's title "arguments" and added it to the Macro properties section.

I'll work to get this rectified.

Kind Regards
Natalie

You can name these files `whatever_you_want.yml` and nest them arbitrarily deep in sub-folders.

<File name='macros/<filename>.yml'>

```yml
version: 2

macros:
  - name: <macro name>
    [description](/reference/resource-properties/description): <markdown_string>
    [docs](/reference/resource-configs/docs):
      show: true | false
    [meta](/reference/resource-configs/meta): {<dictionary>}
    arguments:
      - name: <arg name>
        [type](/reference/resource-properties/arguments): <string>
        [description](/reference/resource-properties/description): <markdown_string>
      - ... # declare properties of additional arguments

  - name: ... # declare properties of additional macros

```

</File>

## type

<File name='macros/<filename>.yml'>

```yml
version: 2

macros:
  - name: <macro name>
    arguments:
      - name: <arg name>
        type: <string>

```

</File>

## Definition

> The data type of your argument. Note that this is only used for documentation purposes — there are no restrictions on the values you can use here.

Reviewer: It's confusing to have two definitions on the page. I think you can remove this title and move these two sentences up right after the  `## type` and before line 67.

---
title: arguments
sidebar_label: "arguments"
id: arguments
---

import MacroArgsNote from '/snippets/_validate-macro-args.md';

<File name='macros/<filename>.yml'>

```yml

version: 2

macros:
  - name: <macro name>
    arguments:
      - name: <arg name>
        type: <string>
        description: <markdown_string>

```

</File>

## Definition

> The `arguments` property is used to define the parameters that a macro can accept. Each argument can have a `name`, `type`, and `description`. 

Reviewer: ```suggestion
The `arguments` property is used to define the parameters that a macro can accept. Each argument can have a `name`, `type`, and `description`.  You add `arguments` to a [macro property](/reference/macro-properties), which helps in documenting the macro and understanding what inputs it requires.
```

> The `arguments` property is used to define the parameters that a macro can accept. Each argument can have a `name`, `type`, and `description`. This helps in documenting the macro and understanding what inputs it requires.

Reviewer: ```suggestion
The `arguments` property is used to define the parameters that a macro can accept. Each argument can have a `name`, `type`, and `description`. 
You add `arguments` to a [macro property](/reference/macro-properties), which helps in documenting the macro and understanding what inputs it requires.
```

> You add `arguments` to a [macro property](/reference/macro-properties), which helps in documenting the macro and understanding what inputs it requires.

Reviewer: ```suggestion
```

> You can define `arguments` in a [`properties.yml`](/reference/macro-properties) file. You can name this file anything and it can placed in the [`macro-paths`](/reference/project-configs/macro-paths) directory (defaults to `["macros"]`). 

Reviewer: ```suggestion
You add `arguments` to a [macro property](/reference/macro-properties). 
```

> You can define `arguments` in a [`properties.yml`](/reference/resource-properties/arguments#macro-properties) file. You can name this file anything and it can placed in the [`macro-paths`](/reference/project-configs/macro-paths) directory (defaults to `["macros"]`). 

Reviewer: ```suggestion
You can define `arguments` in a [`properties.yml`](/reference/macro-properties) file. You can name this file anything and it can placed in the [`macro-paths`](/reference/project-configs/macro-paths) directory (defaults to `["macros"]`). 
```

---
title: arguments
sidebar_label: "arguments"
id: arguments
---

import MacroArgsNote from '/snippets/_validate-macro-args.md';

<File name='macros/<filename>.yml'>

```yml

version: 2

macros:
  - name: <macro name>
    arguments:
      - name: <arg name>
        type: <string>
        description: <markdown_string>

```

</File>

## Definition

The `arguments` property is used to define the parameters that a macro can accept. Each argument can have a `name`, `type`, and `description`. 
You add `arguments` to a [macro property](/reference/macro-properties), which helps in documenting the macro and understanding what inputs it requires.

> You add `arguments` to a [macro property](/reference/macro-properties). 

Reviewer: ```suggestion
```

---
title: arguments
sidebar_label: "arguments"
id: arguments
---

import MacroArgsNote from '/snippets/_validate-macro-args.md';

<File name='macros/<filename>.yml'>

```yml

version: 2

macros:
  - name: <macro name>
    arguments:
      - name: <arg name>
        type: <string>
        description: <markdown_string>

```

</File>

## Definition

The `arguments` property is used to define the parameters that a macro can accept. Each argument can have a `name`, `type`, and `description`. This helps in documenting the macro and understanding what inputs it requires.

> You can use a [behavior flag](/reference/global-configs/behavior-changes#behavior-change-flags) in v1.10. You can define `arguments` in a [`properties.yml`](/reference/resource-properties/arguments#macro-properties) file but they are not enforced by dbt. You can name this file anything and it can placed in the [`macro-paths`](/reference/project-configs/macro-paths) directory (defaults to `["macros"]`).

Reviewer 3: ```suggestion
You can define `arguments` in a [`properties.yml`](/reference/resource-properties/arguments#macro-properties) file. You can name this file anything and it can placed in the [`macro-paths`](/reference/project-configs/macro-paths) directory (defaults to `["macros"]`). 

You can validate your macro `arguments` using the [`validate_macro_args`](/reference/global-configs/behavior-changes#macro-argument-validation) flag. 
```

Reviewer 3: > You can name this file anything and it can placed in the [`macro-paths`](/reference/project-configs/macro-paths) directory (defaults to `["macros"]`).

this feels relevant to general "macro properties" not to `arguments` specifically

Reviewer 3: <img width="927" alt="Screenshot 2025-04-02 at 5 36 11 PM" src="https://github.com/user-attachments/assets/0f11d5af-ac60-4ea2-9350-aecd1b4e1179" />
repetitive of:

- If the flag is set to `False` (default), dbt will continue to permit any value for `type` and `name`.

> - If flag is set to `True` (opt-in), dbt will raise a warning if the argument names you've added in YAML don't match the argument names you have in your macro.

Reviewer 3: ```suggestion
- If flag is set to `True` (opt-in), dbt will raise a warning if the argument names you've added in YAML don't match the argument names you have in your macro or if the argument types aren't valid according to the [supported types](/reference/global-configs/behavior-changes#supported-types).
```

> - If no argument names are documented in YAML, dbt will infer them based on what you have in the macro and include them in the [manifest.json](/reference/artifacts/manifest-json) file. There will be no changes to the manifest and no change to the schema. 

Reviewer 3: ```suggestion

If no argument names are documented in YAML, dbt will infer them based on what you have in the macro and include them in the [manifest.json](/reference/artifacts/manifest-json) file.
```

> ## Macro properties

Reviewer 3: This section "Macro properties" feels like it should be a separate page.

`arguments` is one of the specific macro properties

Reviewer: Agree! Looks like it was mistakenly moved.

Macro properties can be declared in any `properties.yml` file. Macro properties are "special properties" in that you can't configure them in the dbt_project.yml file or using `config()` blocks. Refer to [Configs and properties](/reference/define-properties#which-properties-are-not-also-configs) for more information.

You can name these files `whatever_you_want.yml` and nest them arbitrarily deep in sub-folders.

<File name='macros/<filename>.yml'>

```yml
version: 2

macros:
  - name: <macro name>
    [description](/reference/resource-properties/description): <markdown_string>
    [docs](/reference/resource-configs/docs):
      show: true | false
    [meta](/reference/resource-configs/meta): {<dictionary>}

>     arguments:

Reviewer 3: then we could link to the specific `arguments` page

      - name: <arg name>
        [type](/reference/resource-properties/arguments): <string>
        [description](/reference/resource-properties/description): <markdown_string>
      - ... # declare properties of additional arguments

  - name: ... # declare properties of additional macros

```

</File>

## type

<VersionBlock lastVersion="1.9">

The data type of your argument. This is only used for documentation purposes — there are no restrictions on the values you can use here.

</VersionBlock>
<VersionBlock firstVersion="1.10">

The data type of your argument. Unless you use the [`validate_macro_args`](/reference/global-configs/behavior-changes#macro-argument-validation) flag, `type` is only used for documentation purposes — there are no restrictions on the values you can use here.

</VersionBlock>

<MacroArgsNote />

<File name='macros/<filename>.yml'>

```yml
version: 2

macros:
  - name: <macro name>
    arguments:
      - name: <arg name>
        type: <string>

```

</File>

> ## Supported types for macro argument validation

Reviewer 3: I think we can combine this with the above section on `type`? Or make it a subsection of `type`?

From dbt Core v1.10, when you use the [`validate_macro_args`](/reference/global-configs/behavior-changes#macro-argument-validation) flag, dbt supports the following types for macro arguments:

- `string` or `str`
- `boolean` or `bool`
- `integer` or `int`
- `float`
- `any`
- `list[<Type>]`, for example, `list[string]`
- `dict[<Type>, <Type>]`, for example, `dict[str, list[int]]`
- `optional[<Type>]`, for example, `optional[integer]`
- [`relation`](/reference/dbt-classes#relation)
- [`column`](/reference/dbt-classes#column)

Note that the types follow a Python-like style but are used for documentation and validation only. They are not Python types.

## Examples

### Document a macro

<File name='macros/cents_to_dollars.sql'>

```sql
{% macro cents_to_dollars(column_name, scale=2) %}
    ({{ column_name }} / 100)::numeric(16, {{ scale }})
{% endmacro %}

```

</File>

<File name='macros/cents_to_dollars.yml'>

```yml
version: 2

macros:
  - name: cents_to_dollars
    arguments:
      - name: column_name

>         type: column name or expression

Reviewer 3: ```suggestion
        type: column
```

updating example to use one of the listed accepted `type`s

>         description: "The name of a column, or an expression — anything that can be `select`-ed as a column"

Reviewer 3: ```suggestion
        description: "The name of a column"
```

> 

Reviewer 3: ```suggestion
```
